File: 0208_implement_trie/implement_trie.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class TrieNode
{
public:
    unordered_map<char, TrieNode *> children;
    bool is_word = false;
};

class Trie
{
public:
    void insert(string word);
    bool search(string word);
    bool starts_with(string prefix);

private:
    TrieNode root;
};

/**
 * Inserts a word into the trie.
 */
void Trie::insert(string word)
{
    TrieNode *curr = &root;
    for (auto c : word)
    {
        if (!curr->children.count(c))
            curr->children[c] = new TrieNode();
        curr = curr->children[c];
    }
    curr->is_word = true;
}

/**
 * Determines if word is in the trie.
 */
bool Trie::search(string word)
{
    TrieNode *curr = &root;
    for (auto c : word)
    {
        if (!curr->children.count(c))
            return false;
        curr = curr->children[c];
    }
    return curr->is_word;
}

/**
 * Determines if there is any word in the trie that starts with the
 * given prefix.
 */
bool Trie::starts_with(string prefix)
{
    TrieNode *curr = &root;
    for (auto c : prefix)
    {
        if (!curr->children.count(c))
            return false;
        curr = curr->children[c];
    }
    return true;
}

File: 0208_implement_trie/implement_trie.py
class TrieNode:
    def __init__(self):
        """
        TrieNode constructor.
        """
        self.children = {}
        self.end_of_word = False


class Trie:
    def __init__(self):
        """
        Trie constructor.
        """
        self.root = TrieNode()

    def insert(self, word: str):
        """
        Inserts a word into the trie.
        """
        curr = self.root
        for char in word:
            if char not in curr.children:
                curr.children[char] = TrieNode()
            curr = curr.children.get(char)
        curr.end_of_word = True

    def search(self, word: str) -> bool:
        """
        Determines if word is in the trie.
        """
        curr = self.root
        for char in word:
            if char not in curr.children:
                return False
            curr = curr.children.get(char)
        return curr.end_of_word

    def starts_with(self, prefix: str) -> bool:
        """
        Determines if there is any word in the trie that starts with the
        given prefix.
        """
        curr = self.root
        for char in prefix:
            if char not in curr.children:
                return False
            curr = curr.children.get(char)
        return True
